MazerPlatformer/GameObject.cs

```diff
@@ -162,57 +167,71 @@ public Component AddComponent(Component.ComponentType type, object value, string
+                {
+                    var oldValue = found.Value;
+                    found.Value = newValue;
 
-            OnGameObjectComponentChanged?.Invoke(this, found.Id, found.Type, oldValue, newValue);
-            return true;
-        }
+                    OnGameObjectComponentChanged?.Invoke(this, found.Id, found.Type, oldValue, newValue);
+                    return newValue;
+                });
 
         public void AddState(State state) => States.Add(state);
 
         #region Diganostics
 
         // Draw the centre point of the object
-        protected void DrawCentrePoint(SpriteBatch spriteBatch) => DoIf(Diganostics.DrawCentrePoint, ()=> spriteBatch.DrawCircle(Centre, 2, 16, Color.Red, 3f));
-        
+        protected Either<IFailure, Unit> DrawCentrePoint(SpriteBatch spriteBatch)
+            => EnsureIf(Diganostics.DrawCentrePoint,
+                () =>  spriteBatch.DrawCircle(Centre, 2, 16, Color.Red, 3f))
+                .IgnoreFailure(Nothing);
+
         // Draw the max point (lower right point)
-        protected void DrawMaxPoint(SpriteBatch spriteBatch) => DoIf(Diganostics.DrawMaxPoint, ()=> spriteBatch.DrawCircle(MaxPoint, 2, 8, Color.Yellow, 3f));
+        protected Either<IFailure, Unit> DrawMaxPoint(SpriteBatch spriteBatch) 
+            => EnsureIf(Diganostics.DrawMaxPoint, () => spriteBatch.DrawCircle(MaxPoint, 2, 8, Color.Yellow, 3f))
+                .IgnoreFailure(Nothing);
 
         // Draw the bounding box
-        protected void DrawGameObjectBoundingBox(SpriteBatch spriteBatch) => DoIf(Diganostics.DrawGameObjectBounds, ()=> spriteBatch.DrawRectangle(_boundingBox.ToRectangle(), Color.Lime, 1.5f));
+        protected Either<IFailure, Unit> DrawGameObjectBoundingBox(SpriteBatch spriteBatch) 
+            => EnsureIf(Diganostics.DrawGameObjectBounds, () => spriteBatch.DrawRectangle(_boundingBox.ToRectangle(), Color.Lime, 1.5f))
+                .IgnoreFailure(Nothing);
 
         // Draw the bounding sphere
-        protected void DrawGameObjectBoundingSphere(SpriteBatch spriteBatch) => DoIf(Diganostics.DrawGameObjectBounds, () => spriteBatch.DrawCircle(_centre, BoundingSphere.Radius, 8, Color.Aqua));
+        protected Either<IFailure, Unit> DrawGameObjectBoundingSphere(SpriteBatch spriteBatch) => EnsureIf(Diganostics.DrawGameObjectBounds, () 
+            => Ensure(() => spriteBatch.DrawCircle(_centre, BoundingSphere.Radius, 8, Color.Aqua)))
+            .IgnoreFailure(Nothing);
 
         // Draw all the diagnostics together
-        protected void DrawObjectDiagnostics(SpriteBatch spriteBatch)
-        {
-            DrawCentrePoint(spriteBatch);
-            DrawMaxPoint(spriteBatch);
-            DrawGameObjectBoundingBox(spriteBatch);
-            DrawGameObjectBoundingSphere(spriteBatch);
-        }
+        protected Either<IFailure, Unit> DrawObjectDiagnostics(SpriteBatch spriteBatch) =>
+            DrawCentrePoint(spriteBatch)
+                .Bind(unit => DrawMaxPoint(spriteBatch))
```

Comment:
It failed here before when I was returning a ConditionNotMet Failure and it cancelled running all the subsequent diagnostics drawing routines. Solution is to IgnoreFailure() on Eithers that you don't care about the failures and don't want the failures stopping the entire pipeline